### ISSUE TITLE
Replace go-etcd with etcd/client

### ISSUE THREAD

**Issue by bcwaldon** [kubernetes/kubernetes#11962]
The go-etcd client library (github.com/coreos/go-etcd) is deprecated in favor of the client library that lives directly in the etcd repo (github.com/coreos/etcd/client). This new library is receiving a lot more love from the core etcd development team, and is the current blessed set of golang etcd bindings.

**Comment by bprashanth** [comment]
Does it reuse tcp connections or do we have to port our patch? https://github.com/GoogleCloudPlatform/kubernetes/tree/master/third_party/forked/coreos/go-etcd

**Comment by bcwaldon** [comment]
I can't tell you for sure, but I do see that the client sets a default KeepAlive value on a customer net.Dialer: https://github.com/coreos/etcd/blob/master/client/client.go#L43. Maybe the etcd team can chime in here.

**Comment by bprashanth** [comment]
@lavalamp @wojtek-t

**Comment by yichengq** [comment]
> Does it reuse tcp/http connections or do we have to port our patch?

@bprashanth You still need to set `MaxIdleConnsPerHost` for `Transport` in `client.Config`. But you don't need to fork anything, and you could set the transport used when creating the client.

The benefits I could see:
1. cleaner API. keysAPI has only 7 calls now(http://godoc.org/github.com/coreos/etcd/client#KeysAPI), and it is easily extensible.
2. It support members and auth API.
3. it is thread-safe.

Possible drawbacks:
1. It doesn't fasten its unmarshal function yet. Here is the issue to track it: https://github.com/coreos/etcd/issues/3193

**Comment by wojtek-t** [comment]
I think that we should wait until https://github.com/coreos/etcd/issues/3193 is fixed (it doesn't seem to be a lot of work). Once this is fixed we should be able to migrate to the new library relatively quickly.

**Comment by timothysc** [comment]
2 questions: 

1. Do we have a time-horizon on this integration? 
2. Given that the change could be destabilizing, do we have a vetting plan, e.g.  (-devel) vs. 1.1? The previous client had a lot of cycles under its belt prior to 1.0, and I would feel far more comfortable putting such changes into a -devel series. 

cc @smarterclayton @bgrant0607 @dchen1107

**Comment by bgrant0607** [comment]
What problems would this solve? What performance gain could be expected? What bugs would it fix? Will the existing library be deprecated or deleted at some point?

If we need to make the change in the 1.1 timeframe, earlier would be better. If someone could do that and run all the e2e tests that would help us assess the stability risk.

cc @lavalamp

**Comment by timothysc** [comment]
@bgrant0607 The key issue of interest is the thread safety issue outlined in #7566 . 
This effects HA, and read-perf in a HA env.

**Comment by bgrant0607** [comment]
Someone just needs to try it and run all of our e2e tests and benchmarks.

**Comment by timothysc** [comment]
@yichengq do you have an ETA?

**Comment by yichengq** [comment]
@timothysc I plan to finish the left performance issue coreos/etcd#3193 in the next week.

**Comment by wojtek-t** [comment]
I can try to pick it up then.

**Comment by timothysc** [comment]
Per our discussion the client is available: 
https://github.com/coreos/etcd/tree/master/client + godoc:
https://godoc.org/github.com/coreos/etcd/client

@wojtek-t is isolating the dependencies which appear to be in a lot of locations, including a forked client in server.go.?.?.? 

$ grep -r "go-etcd" * | wc -l 
39

I'm starting from the helper and updating.

**Comment by wojtek-t** [comment]
Thanks @timothysc 

One PR slightly reducing dependencies on go-etcd is already LGTMed: 
https://github.com/kubernetes/kubernetes/pull/13509

I have another one pretty close - should send it out for review today and will post its number here.

**Comment by wojtek-t** [comment]
But I think that we will need to convert few things anyway. These are:
- the whole pkg/storage
- references to it under pkt/tool
- few reverences to the client itself under test/integration, cmd/ and cluster/addons

**Comment by wojtek-t** [comment]
No I didn't touch server.go

However, once #13584 is merged it will be much better:

$ git grep "go-etcd" | wc -l
23

(And changing half of them should be trivial)

**Comment by timothysc** [comment]
Deps list for those watching: https://github.com/coreos/etcd/issues/3471

**Comment by timothysc** [comment]
Working through the tests... Should have PR early next week. https://github.com/timothysc/kubernetes/tree/new-etcd-client

**Comment by timothysc** [comment]
@wojtek-t fwiw there are a series of follow on-s around this.  In part b/c I'm not a fan of some of the abstractions in the code, and I believe we should try to push all dependencies behind the storage interface.  

In general, I think pImpl/envelop-letter idioms need to be applied more thoroughly in the code to prevent dependency sprawl.

**Comment by glerchundi** [comment]
+1, It will be cool to have this merged.

I run into a weird issue that probably is related to the client library, take a look to this project: https://github.com/glerchundi/etcd-go-clients-test. It compares empirically an access to a concrete etcd key using a load balancer.

@yichengq do you know what could be happening in there?

You could find complete issue doc. here: #14977.

**Comment by philips** [comment]
@glerchundi that URL doesn't work.

**Comment by glerchundi** [comment]
Yep, I removed it because of this:
http://stackoverflow.com/questions/32912699/unable-to-setup-kubernetes-against-tls-secured-etcd.

@davidopp opened an issue that probably will prevent people from falling back into the same problem https://github.com/kubernetes/kubernetes/issues/15056.

**Comment by wojtek-t** [comment]
@jonboulle FYI

**Comment by timothysc** [comment]
just an FYI, we are breaking up the commits on this as noted above.

**Comment by timothysc** [comment]
Just to give a complete update.. 
- 1st commit around contexts is in. 
- We've agreed upon removing the FakeClient and stading up etcd per unit test that needs it.  Currently block on godep issues 
- Code exists to transition to new client, but there is some runtime error that is masked by the FakeClient and only visible by the integration tests.

**Comment by smarterclayton** [comment]
I'd like more context behind the second bullet - that doesn't make a lot of
sense to me to require a database to be running to run the bulk of our unit
tests that touch the storage path.

**Comment by timothysc** [comment]
The SIG agreed that the FakeClient doesn't make much sense going forwards as the semantics change with both the current mod, as well as the upcoming etcd3 mods.  We are basically fixing the fake client to match the shift, while it still masks behavioral differences between clients. 

Thus standing up etcd programmatically is relatively low cost and can be done on an as needed basis.   

https://github.com/kubernetes/kubernetes/pull/15392

**Comment by smarterclayton** [comment]
Can you explain why FakeClient doesn't make sense?  Emulating deep etcd
behavior I agree is not productive.  But higher level tests having to start
etcd concerns me more.

On Wed, Oct 14, 2015 at 1:17 PM, Timothy St. Clair <notifications@github.com

**Comment by timothysc** [comment]
> Can you explain why FakeClient doesn't make sense?  Emulating deep etcd behavior I agree is not productive.

The updated client libraries change both the api and behaviour.  On those modifications, it forces a shift in the FakeClient to match on both fronts.  So shift now + shift later ... 

> But higher level tests having to start etcd concerns me more.

We counted ~15 files, and not all tests within those files.  In general, the code has not properly abstracted the storage layer, and that concerns me even more.  

Removing the FakeClient ensures that we are testing reality as this changes.  Ideally we hide behind the storage layer for all of this.

**Comment by smarterclayton** [comment]
Ok, thanks for the clarification.

On Wed, Oct 14, 2015 at 3:12 PM, Timothy St. Clair <notifications@github.com

**Comment by timothysc** [comment]
For update, code needed to replace fakeClient is now I'm in... 
Currently in flight on replacement, and will remove /tool dependencies and shift to storage as we go.

**Comment by timothysc** [comment]
Storage layer is nearly completely isolated only a couple of minor changes needed now.

**Comment by wojtek-t** [comment]
There are at least two places where this should be relatively simple to get rid of the dependency:
- contrib/mesos
- pkg/master/master_test.go
I will try to do at least second one tomorrow.

**Comment by wojtek-t** [comment]
Also - I would like to eliminated tools.EtcdClient interface completely,

**Comment by timothysc** [comment]
agreed

**Comment by liggitt** [comment]
any update on this? I'm interested in being able to plumb a quorum read in a couple particular cases, and only the new client exposes that option

**Comment by timothysc** [comment]
I'll get the PR in this week.

**Comment by yuvipanda** [comment]
/me registers interest in this since that'll probably allow easier usage of etcd ACLs

**Comment by timothysc** [comment]
@wojtek-t we can finally close this one now.

**Comment by wojtek-t** [comment]
I think there is two more actions to do here
- replace go-etcd with new etcd client in cluster/addons/dns/kube2sky
- removed go-etcd dependency completely.

**Comment by timothysc** [comment]
There is a separate issue for that.  #18791

**Comment by wojtek-t** [comment]
OK - then I guess we can close this one.
